parsec/data.c

```diff
@@ -559,3 +562,28 @@ parsec_data_destroy( parsec_data_t *data )
 #endif
     PARSEC_OBJ_RELEASE(data);
 }
+
+void
+parsec_data_discard( parsec_data_t *data )
+{
+
+    /* first release the reference the application held */
+    PARSEC_OBJ_RELEASE(data);
+
+    /* second, mark the host copy as discarded */
+    parsec_data_copy_t *cpu_copy = data->device_copies[0];
+    if (NULL != cpu_copy) {
+        cpu_copy->flags = PARSEC_DATA_FLAG_DISCARDED;
```

Comment:
Three issues here:
- the flag should be set with `|=`
- you should only release the `data_t` is the `cpu_copy` is valid. You do a test against NULL, you should use it.
- this function is not thread-safe. This could not be an issue here, but I see it used by the GPU code without protection either, opening the code to race-conditions while manipulating these discarded copies/

Reply:
Good catch. I fixed setting the flag, added a lock to protect the check for the change to the device copy, and only now the data_t is only released twice if there is a device copy.